### ISSUE TITLE
Gutenberg should prompt users to create ALT text for images

### ISSUE THREAD

**Issue by zeldman** [WordPress/gutenberg#16909]
An easy #a11y win would be to make ALT text mandatory for non-decorative images, by, at the very least, putting up a warning flag if someone neglects to create alt text. 

**WHAT SHOULD HAPPEN**

When the user adds an image, Gutenberg should prompt them to create an ALT text, and flag them with a warning if they neglect to do so.

**WHAT DOES HAPPEN** 

Nothing.

**RELATED ISSUES**

Currently there is an Issue on helping the user avoid creating INAPPROPRIATE alt text:

https://github.com/WordPress/gutenberg/issues/5205

That’s fine as far as it goes, but it ignores the bigger problem that most users really will not bother creating alt text unless we programmatically urge them to do so.

Alt text is a bare minimum accessibility effort; it’s literally the least a content creator can do. Let’s help them make sure to do it!

Hat tip: Kjell Riegstad for sharing this issue with me and Mark Uraine for guidance on sharing this request.

**Comment by karmatosed** [comment]
A possible related issue is: #12202

**Comment by ellatrix** [comment]
Cc @afercia.

**Comment by afercia** [comment]
This is something that would greatly help in producing accessible content and I'd totally second it.

I guess the technical challenge would be implementing a mechanism to determine whether an image is _decorative_ or not.

**Single, linked, images**

The simplest case is probably the one @joedolson [mentioned on 5205](https://github.com/WordPress/gutenberg/issues/5205#issuecomment-430317236): a linked image with no other content. The image block would make this relatively simple, as both the `alt` attribute and the image link are block settings:

Gutenberg image block on WordPress 5.2.2:

![Screenshot 2019-08-06 at 10 44 54](https://user-images.githubusercontent.com/1682452/62526822-4c589b00-b83a-11e9-8bc1-eeb5cf973c61.jpg)

Worth noting the interface has changed in the Gutenberg plugin: the alt text description includes a link to [the W3C alt decision tree tutorial](https://www.w3.org/WAI/tutorials/images/decision-tree/). The UI to link the image has changed as well and it's now within the block toolbar:

![Screenshot 2019-08-06 at 10 46 19](https://user-images.githubusercontent.com/1682452/62527461-765e8d00-b83b-11e9-8766-8a8545752deb.jpg)

I'd dare say Gutenberg shouldn't even allow to save a linked image which has an empty alt attribute, as that actually produces the equivalent of an empty link:

`<a href="https://somewhere"><img src="..." alt="" /></a>`

**Single, non-linked, images**
I'm not sure how the relevance of non-linked images can be determined programmatically. When an image is purely decorative it should have an empty alt attribute `alt=""`. Only when an image adds relevant information and "contributes meaning to the current page or context" it should be described following the guidance provided on https://www.w3.org/WAI/tutorials/images/decision-tree/. Depending on the complexity of the image, an `alt` attribute might be not appropriate and the description should be provided elsewhere on the page.

Regardless, Gutenberg should be smart enough to understand when an image is non-decorative, which implies some sort of image recognition and text analysis mechanisms.

**Images in the Classic block and Custom HTML block**
This is probably the most complicated case. For example, a link might contain an image _and_ some text. It the text already describes the link purpose, [the image should use an empty alt attribute](https://www.w3.org/TR/WCAG20-TECHS/H2.html). In this case Gutenberg should check the actual content of the link first, and then use image recognition and text analysis. 

Allowing free-form HTML exposes to all sort of cases. Not sure they can be all analyzed programmatically in a proper way. 

**Images in a gallery**
They probably need to always have an `alt` attribute, whether they're linked or not. Images in a gallery determine the actual content of the gallery. Gallery images without an `alt` attribute defeat the purpose of a gallery.

**Images from other blocks and custom blocks**
For example: the "Media & text" block, and blocks provided by plugins (the "Cover" block uses a background image). Same as above for non-linked and linked images.

**Images from embeds**
For example: the Instagram block, the Flickr block, etc. I'm afraid there's not so much Gutenberg can do in these cases. The embed content is the one returned by the embed provider. For example, Instagram doesn't even use an empty alt attribute so screen readers fallback to the image filename:

<img width="847" alt="Screenshot 2019-08-06 at 11 41 36" src="https://user-images.githubusercontent.com/1682452/62530143-48c81280-b840-11e9-831e-04c441b7e38e.png">

(Instagram doesn't even provide a `title` attribute on the `iframe` element)

**Comment by iandunn** [comment]
I think this is a great idea, it'll help educate users about writing accessible content, and make a significant impact for readers.

> I'd dare say Gutenberg shouldn't even allow to save a linked image which has an empty alt attribute

I really like the idea of showing a warning, similar to what we do for color contrast. I'd be very, very cautious when considering making it mandatory, though; that feels very heavy-handed and likely to backfire, annoying users to the point where they resent it and are less likely to want to write accessible content.

[Honey, not vinegar...](https://www.dictionary.com/browse/you-can-catch-more-flies-with-honey-than-with-vinegar)

> Gutenberg should be smart enough to understand when an image is non-decorative, which implies some sort of image recognition and text analysis mechanisms.

I'd also be leery of trying to programmatically detect whether or not an image is decorative. That seems overly complex and prone to error, but allowing users to make the decision lets them retain control, and is an opportunity for them to learn about writing accessible content.

**Comment by joedolson** [comment]
The most definitive way to detect whether an image is not decorative is the presence of text in the image - but even that can backfire, if it's an image that happens to have a sign in the background that's legible but not actually relevant. 

A linked image with no alt attribute, however, is 100% always an accessibility error, and should always be flagged as such.

The most legitimate case where I think it should be permissible to insert an image where there's an image that needs an appropriate alt attribute would be to allow sight impaired users to insert images with the expectation that somebody else will be able to later add appropriate alt attributes. 

I'm in favor of adding strong and clear error flags, but feel that any situation that actively prevents actions has the potential to create user interaction barriers.

**Comment by ryanboren** [comment]
> “Blind users also lamented that the alt-text options on Instagram and Facebook are hard to find for sighted users if they don’t know to look for them as they upload a photo.”

Source: https://slate.com/technology/2019/11/facebook-blind-users-no-accessibility.html

I'm looking at Gutenberg in iOS Safari on an iPhone. To get to alt text for an image in the editor: three dot menu -> Show Block Settings -> scroll down to alt text in Image Settings. It's buried and ignorable. Further, the image being described does not show on screen with the alt text editing box, making  describing the image difficult.

**Comment by cbirdsong** [comment]
Flagging alt text-less linked images and automatically detecting decorative images are worth investigating but they feel like separate issues. I think the core here is this:

> Alt text is a bare minimum accessibility effort; it’s literally the least a content creator can do. Let’s help them make sure to do it!

Users remembering to add alt text is a huge problem I see day to day, even from people who know it's necessary. It's easy to miss, because alt text is not in the basic user flow for adding an image. It's relegated to the sidebar, which is [described this way in the block editor handbook](https://developer.wordpress.org/block-editor/designers/block-design/): 

> ### The Settings Sidebar should only be used for advanced, tertiary controls 
> 
> The Settings Sidebar is not visible by default on a small / mobile screen, and may also be collapsed in a desktop view. Therefore, it should not be relied on for anything that is necessary for the basic operation of the block. Pick good defaults, make important actions available in the block toolbar, and think of the Settings Sidebar as something that most users should not need to open.

Alt text is not a tertiary function. Adding or omitting it should be a core part of using the image block. My suggested solution: 

When an image inserted, the default state becomes a prompt for alt text floating over a tinted version of the image, reminiscent of the cover block. The user can dismiss the prompt if the image is decorative or leave it alone as a reminder to come back later. A toggle to show/hide it should also be made available in the image block's toolbar. Here's a quick _extremely rough_ mockup made in the inspector:

![Screenshot of the block editor containing a tinted image with a prompt for alt text on top](https://user-images.githubusercontent.com/1672206/73857704-e8b92580-47fc-11ea-8b0c-d08fbe350b32.png)

No warning or blocking. Just a simple prompt, right in line with the rest of the content.

**Comment by joedolson** [comment]
I think that having an alt text field as the default overlay when editing any inserted image without an alt attribute is a nice idea. Easy to find, image is immediate, etc.

**Comment by carolinan** [comment]
I can picture this overlay becoming annoying for users and that they would only hit the close button.
After doing that 5 times, it would be come more than an annoyance.

**Comment by cbirdsong** [comment]
I hadn't imagined it being implemented as a fullscreen takeover – it should just sit over the image in the editor acting as a reminder to enter alt text, but not interrupting the editing experience. 

The core problem is that adding alt text should be part of the normal user flow instead of being buried in a sidebar/submenu.

**Comment by joedolson** [comment]
Yes, I hadn't seen that as something that was a takeover; the interaction would completely be optional - dismiss, complete, ignore. 

I would consider that the placement possibly shouldn't be the center of the image, so it doesn't block the view of the image that needs to be described too much, but this otherwise seems like a totally reasonable interface.

**Comment by wkhayrattee** [comment]
hey everyone, the last comment was a year ago, wanted to know if there has been any update to address the aim of this thread?

**Comment by cbirdsong** [comment]
Bumping this issue up to point out how Mastodon handles this after seeing it [praised by Matt Haughey, who wrote](https://a.wholelottanothing.org/2022/11/08/clever-thoughtful-software-design/):

> After 15 years of using Twitter, I only recently got an option that I had to turn on deep in my settings to give me the ability to write alt descriptions on images. Also, the process of using the feature isn’t smooth, you have to remember to tap images and write one, or when you publish you get a giant popup saying you forgot to include the text and would you like to add it before publishing.
> 
> […] 
> 
> ![iPhone screenshot of an image being added to a post on Mastodon. A text input area with placeholder text reading "Describe the photo for the visually impaired" is (barely) visible on top of the image.](https://user-images.githubusercontent.com/1672206/201360028-0b129644-5526-48f1-80c0-e49a571956e4.png)
> 
> At Mastodon, you upload an image to a new post and there’s some text that gently hints how to add a description and why you would do it. Adding it is easy, you tap the image and start typing while you’re looking at it. And if you forget, the software doesn’t punish you.

Mastodon’s design is far from perfect[^1], but it’s exactly the direction I had in mind: An on-by-default text input that offers the user a reminder if they’ve not entered an image description without nagging or penalizing them.

[^1]: I very much doubt that placeholder text passes a contrast test, and it would be difficult to add more than a sentence or two without it getting very awkward.

**Comment by mediaformat** [comment]
Bumping, if no one else has started work on this I'm willing to give it a go.

**Comment by joedolson** [comment]
Please do!

**Comment by mediaformat** [comment]
Coming back with notes from when this was last discussed on slack

Proposed Scope:
- Image block

Design:
- A visual indication of missing alt text (in the editor)
  - Cross browser / device
- A checkbox for indicating that the image is decorative
- ~~A pre-publish check: ensuring image either has alt text, or has checkbox declaring the image to be decorative~~. It looks like the [PluginPrePublishPanel](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-pre-publish-panel/) is not be intended for Core use (based on its name, and lack of use in elsewhere in core). So I'm hesitant to add this to the PR I've started.

**Comment by joedolson** [comment]
@troychaplin This issue would correlate well with a lot of the work you've been doing with your block accessibility checker.
